cpp/src/Ice/Reference.cpp

```diff
@@ -1572,137 +1572,85 @@ IceInternal::RoutableReference::createConnectionAsync(
         return;
     }
 
-    //
-    // Finally, create the connection.
-    //
     OutgoingConnectionFactoryPtr factory = getInstance()->outgoingConnectionFactory();
-    if(getCacheConnection() || endpoints.size() == 1)
+    auto self = static_pointer_cast<RoutableReference>(const_cast<RoutableReference*>(this)->shared_from_this());
+
+    auto createConnectionResponse =
+        [routerInfo = _routerInfo, response = std::move(response)](Ice::ConnectionIPtr connection, bool compress)
     {
-        class CB1 final : public OutgoingConnectionFactory::CreateConnectionCallback
+        // If we have a router, set the object adapter for this router (if any) to the new connection, so that
+        // callbacks from the router can be received over this new connection.
+        if (routerInfo && routerInfo->getAdapter())
         {
-        public:
-
-            void setConnection(const Ice::ConnectionIPtr& connection, bool compress) final
-            {
-                //
-                // If we have a router, set the object adapter for this router
-                // (if any) to the new connection, so that callbacks from the
-                // router can be received over this new connection.
-                //
-                if(_routerInfo && _routerInfo->getAdapter())
-                {
-                    connection->setAdapter(_routerInfo->getAdapter());
-                }
-                _response(connection, compress);
-            }
-
-            void setException(std::exception_ptr ex) final
-            {
-                _exception(ex);
-            }
-
-            CB1(
-                const RouterInfoPtr& routerInfo,
-                function<void(ConnectionIPtr, bool)> response,
-                function<void(std::exception_ptr)> exception) :
-                _routerInfo(routerInfo), _response(std::move(response)), _exception(std::move(exception))
-            {
-            }
-
-        private:
-
-            const RouterInfoPtr _routerInfo;
-            const function<void(ConnectionIPtr, bool)> _response;
-            const function<void(std::exception_ptr)> _exception;
-        };
+            connection->setAdapter(routerInfo->getAdapter());
+        }
+        response(std::move(connection), compress);
+    };
 
-        //
-        // Get an existing connection or create one if there's no
-        // existing connection to one of the given endpoints.
-        //
-        factory->create(endpoints, false, getEndpointSelection(), make_shared<CB1>(_routerInfo, response, exception));
-        return;
+    if(getCacheConnection() || endpoints.size() == 1)
+    {
+        // Get an existing connection or create one if there's no existing connection to one of the given endpoints.
```

Comment:
I find this code too complicated, with a lambda that captures a response and itself and then needs to clear itself.

There must a clearer way to write all this.